src/gui/dialogs/units_dialog.cpp

```diff
@@ -246,7 +246,16 @@ void units_dialog::dismiss_unit(std::vector<unit_const_ptr>& unit_list, const te
 
 	// If the unit is of level > 1, or is close to advancing, we warn the player about it
 	std::stringstream message;
-	if(u.loyal()) {
+	const auto& traits_list = u.get_traits_list();
+	bool has_loyal_trait = std::find(traits_list.begin(), traits_list.end(), "loyal") != traits_list.end();
+
+	// first case: has no upkeep but is not a loyal unit, and not a level zero
+	if(u.upkeep() == 0 && u.level() >= 1 && !has_loyal_trait) {
+		message << _("This unit requires no upkeep.") << " " << (u.gender() == unit_race::MALE
+		         ? _("Do you really want to dismiss him?")
+		         : _("Do you really want to dismiss her?"));
+	// second case: has no upkeep and is a loyal unit
+	} else if(u.upkeep() == 0 && has_loyal_trait) {
```

Comment:
I think the correct thing to do here is to check `upkeep_raw()` instead of `upkeep()`, because there exists the edge case of a unit who has `upkeep=full` which evaluates to 0 because it is a level 0 unit, and it doesn't make sense to show the warning in that case.

In other words:

```suggestion
	if(u.upkeep_raw() == unit::upkeep_loyal()) {
```